fpga_interchange/arch_pack_clusters.cc

```diff
@@ -0,0 +1,496 @@
+    NPNR_ASSERT(loc_last != Loc());
+    ArcBounds bounds(
+        std::min(loc_root.x, loc_last.x),
+        std::min(loc_root.y, loc_last.y),
+        std::max(loc_root.x, loc_last.x),
+        std::max(loc_root.y, loc_last.y)
+    );
+    return bounds;
+}
+
+Loc Arch::getClusterOffset(const CellInfo *cell) const
+{
+    Loc offset;
+    if (cell->bel != BelId()) {
+        CellInfo *root = getClusterRootCell(cell->cluster);
+        Loc root_loc = getBelLocation(root->bel);
+        Loc cell_loc = getBelLocation(cell->bel);
+        offset.x = cell_loc.x - root_loc.x;
+        offset.y = cell_loc.y - root_loc.y;
+        offset.z = cell_loc.z - root_loc.z;
+    }
+
+    return offset;
+}
+
+bool Arch::isClusterStrict(const CellInfo *cell) const
+{
+    return true;
+}
+
+void dump_chains(const ChipInfoPOD *chip_info, Context *ctx)
+{
+    std::unordered_map<IdString, const BelChainPOD *> bel_chain_prototypes;
+    for (size_t i = 0; i < chip_info->bel_chains.size(); ++i) {
+        const auto &bel_chain = chip_info->bel_chains[i];
+        IdString bel_chain_name(bel_chain.name);
+        bel_chain_prototypes.emplace(bel_chain_name, &bel_chain);
+        log_info("BEL chain '%s' loaded! Parameters:\n", bel_chain_name.c_str(ctx));
+        log_info("  - sites:\n");
+        for (auto site : bel_chain.sites) {
+            IdString site_name(site);
+            log_info("      - %s\n", site_name.c_str(ctx));
+        }
+
+        log_info("  - cells:\n");
+        for (auto cell : bel_chain.cells) {
+            IdString cell_name(cell);
+            log_info("      - %s\n", cell_name.c_str(ctx));
+        }
+
+        log_info("  - pattern:\n");
+        IdString source_cport(bel_chain.source->cell_port);
+        IdString source_bport(bel_chain.source->bel_port);
+        IdString sink_cport(bel_chain.sink->cell_port);
+        IdString sink_bport(bel_chain.sink->bel_port);
+        log_info("      - src %s - %s --> dst %s - %s\n", source_cport.c_str(ctx), source_bport.c_str(ctx), sink_cport.c_str(ctx), sink_bport.c_str(ctx));
+
+        log_info("  - drivers:\n");
+        for (auto &driver : bel_chain.chain_drivers) {
+            log_info("    - in only cells:\n");
+            for (auto cell : driver.in_cells)
+                log_info("      - %s\n", IdString(cell).c_str(ctx));
+            log_info("    - in and out cells:\n");
+            for (auto cell : driver.in_out_cells)
+                log_info("      - %s\n", IdString(cell).c_str(ctx));
+            log_info("    - ports:\n");
+            for (auto port : driver.ports) {
+                log_info("      - %s - %s\n", IdString(port.name).c_str(ctx), IdString(port.bel).c_str(ctx));
+            }
+        }
+    }
+}
+
+void Arch::prepare_cluster(const BelChainPOD *chain)
+{
+    Context *ctx = getCtx();
+    IdString chain_name(chain->name);
+
+    // Get chainable cells
+    std::vector<CellInfo *> chainable_cells;
+    for (auto cell : sorted(ctx->cells)) {
```

Comment:
another note post #718, this will need to be changed to `for (auto &cell : ctx->cells) {` `CellInfo *ci = cell.second.get();`

Reply:
Fixed